klog/klog.c

```diff
@@ -0,0 +1,348 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system kernel
+ *
+ * Kernel log buffer
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Maciej Purski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include HAL
+
+#include "../include/ioctl.h"
+
+#include "../posix/posix.h"
+
+#include "klog.h"
+#include "../proc/threads.h"
+#include "../proc/msg.h"
+#include "../proc/ports.h"
+
+
+#ifndef KLOG_BUFSZ
+#define KLOG_BUFSZ (2 * SIZE_PAGE)
+#endif
+
+
+#define TCGETS 0x405c7401
+
+
+typedef struct klog_reader {
+	void *odata;
+	size_t osize;
+	offs_t ridx;
+	unsigned long rid;
+	unsigned pid;
+	unsigned nonblocking;
+	unsigned waiting;
+	struct klog_reader *prev, *next;
+} klog_reader_t;
+
+
+static struct {
+	char buf[KLOG_BUFSZ];
+	u32 port;
+	offs_t head;
+	offs_t tail;
+	lock_t lock;
+	klog_reader_t *readers;
+} klog_common;
+
+
+static int _fifo_empty(void)
+{
+	return klog_common.tail == klog_common.head;
+}
+
+
+static int _fifo_full(void)
+{
+	return klog_common.tail - klog_common.head == KLOG_BUFSZ;
+}
+
+
+static char _fifo_pop(void)
+{
+	return klog_common.buf[klog_common.head++ % KLOG_BUFSZ];
+}
+
+
+static void _fifo_push(char c)
+{
+	klog_common.buf[klog_common.tail % KLOG_BUFSZ] = c;
+	klog_common.tail++;
+	if (_fifo_full())
+		_fifo_pop();
+}
+
+
+static char _fifo_get(offs_t off)
+{
+	return klog_common.buf[off % KLOG_BUFSZ];
+}
+
+
+void _klog_init(void)
+{
+	hal_memset(&klog_common, 0, sizeof(klog_common));
+	proc_lockInit(&klog_common.lock);
+}
+
+
+static klog_reader_t *klog_readerFind(unsigned pid)
+{
+	klog_reader_t *r = klog_common.readers;
+
+	if (r != NULL) {
+		do {
+			if (r->pid == pid)
+				return r;
+			r = r->next;
+		} while (r != klog_common.readers);
+	}
+
+	return NULL;
+}
+
+
+static int klog_readerAdd(unsigned pid, unsigned nonblocking)
+{
+	klog_reader_t *r;
+
+	if (klog_readerFind(pid) != NULL)
+		return -EINVAL;
+
+	if ((r = vm_kmalloc(sizeof(klog_reader_t))) == NULL)
+		return -ENOMEM;
+
+	hal_memset(r, 0, sizeof(klog_reader_t));
+
+	r->nonblocking = !!nonblocking;
+	r->pid = pid;
+
+	proc_lockSet(&klog_common.lock);
+	r->ridx = klog_common.head;
+	LIST_ADD(&klog_common.readers, r);
+	proc_lockClear(&klog_common.lock);
+
+	return 0;
+}
+
+
+static int _klog_readln(klog_reader_t *r, char *buf, size_t sz)
+{
+	size_t n = 0;
+
+	while (r->ridx < klog_common.tail && n < sz) {
+		buf[n++] = _fifo_get(r->ridx++);
+		if (buf[n - 1] == '\n' || buf[n - 1] == '\0')
+			break;
+	}
+
+	/* Always end with newline */
+	if (n > 0 && buf[n - 1] != '\n') {
+		if (n < sz)
+			buf[n++] = '\n';
+		else
+			buf[n - 1] = '\n';
+	}
```

Comment:
1. Input: `line = "abcd\n" sz = 2` 
`n == 2` and `buf[1] == 'b'`
You overwrite char in `buf[n -1]` with a newline char.
Output:
`buf == "a\n" n == 2`

2. Input: `line="abc\0" sz = 10`
`n == 4 buf[3] == '\0'`
You append '\n' to the oryginal line
Output:
`buf="abc\0\n" n == 5`

Reply:
The first case is desired and that is what I meant. The idea is to *always* end one read with a newline character. So we read as many characters as we can, but always finish with `\n`. However now this overwritten character would be completely lost, so it would be better to update a reader's read index in order to read this character with a next read.

The second case is arguable, but I think it would be better to overwrite this `\0` character.